File: Serial OrderWise/LeetCode 1011 - Capacity To Ship Packages Within D Days/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
        int shipWithinDays(vector<int> &weights, int days)
        {
                int s = INT_MIN, e = 0;
                for (int i : weights)
                {
                        e += i;
                        s = max(i, s);
                }
                while (s < e)
                {
                        int mid = s + (e - s) / 2;
                        int need = 1, curr = 0;
                        for (int i : weights)
                        {
                                if (curr + i > mid)
                                {
                                        need++;
                                        curr = 0;
                                }
                                curr += i;
                        }
                        if (need > days)
                        {
                                s = mid + 1;
                        }
                        else
                        {
                                e = mid;
                        }
                }
                return s;
        }
};

File: Serial OrderWise/LeetCode 1011 - Capacity To Ship Packages Within D Days/Solution.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        s, e = -1, 0
        for i in weights:
            s = max(s, i)
            e += i
        while s < e :
            mid = s + (e-s)//2
            need = 1
            curr = 0
            for i in weights:
                if curr + i > mid:
                    need += 1
                    curr = 0
                curr += i
            if need > days :
                s = mid + 1
            else :
                e = mid
        return s 